scripts/common/kubernetes.sh

```diff
@@ -316,13 +316,26 @@ function spinner_kubernetes_api_stable() {
 }
 
 function kubernetes_drain() {
-    kubectl drain "$1" \
+    if kubernetes_has_remotes ; then
```

Comment:
could you simplify the code here and omit this conditional if you use `enable-deletion` in conjunction with `skip-wait-for-delete-timeout > 0`?

The flag is hard to understand but this post does a good job of explaining

https://github.com/rancher/system-upgrade-controller/discussions/148

> If you would prefer to honor PDB up to a maximum-ish number of seconds, then you should set your plan .spec.drain.skipWaitForDeleteTimeout.

Reply:
Actually I'm finding that when I'm upgrading kubernetes on 4 node cluster I'm still hitting PDBs when rolling the 4th node since the other 3 have just been upgraded. Eventually they do get drained but I still wonder if setting `skip-wait-for-delete-timeout > 0` as a precaution would be best?

```
...
evicting pod longhorn-system/instance-manager-r-4f2ebd0b
evicting pod longhorn-system/instance-manager-e-884b7b96
error when evicting pods/"instance-manager-e-884b7b96" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pods/"instance-manager-r-4f2ebd0b" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod longhorn-system/instance-manager-r-4f2ebd0b
evicting pod longhorn-system/instance-manager-e-884b7b96
error when evicting pods/"instance-manager-e-884b7b96" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pods/"instance-manager-r-4f2ebd0b" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod longhorn-system/instance-manager-e-884b7b96
evicting pod longhorn-system/instance-manager-r-4f2ebd0b
error when evicting pods/"instance-manager-r-4f2ebd0b" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pods/"instance-manager-e-884b7b96" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod longhorn-system/instance-manager-e-884b7b96
evicting pod longhorn-system/instance-manager-r-4f2ebd0b
error when evicting pods/"instance-manager-r-4f2ebd0b" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pods/"instance-manager-e-884b7b96" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod longhorn-system/instance-manager-e-884b7b96
evicting pod longhorn-system/instance-manager-r-4f2ebd0b
pod/instance-manager-r-4f2ebd0b evicted
pod/instance-manager-e-884b7b96 evicted
node/ethanm-migrate-4 drained
```

Reply:
> if you use enable-deletion

What is `enable-deletion`?

I understand what you're saying: we can still respect PDBs (for multi node scenarios) and fall back to deleting the pods using one command. I'll make this change and test it.

Reply:
Sorry I meant `disable-eviction`

Reply:
>skip-wait-for-delete-timeout > 0 as a precaution would be best

What's the benefit of adding this flag? The main benefit I see is that it ensures `kubectl drain` command completes instead of timing out.

Reply:
It will try normal eviction until the wait for delete timeout and then move on to delete. Evict is the normal drain behavior. I haven't looked into the difference between delete and evict. Do you think its ok to default to delete rather than evict?

https://stackoverflow.com/questions/62277852/whats-the-difference-between-pod-deletion-and-pod-eviction

Reply:
Deleting a pod _should_ be fine since they are backed by a controller. As I stated before, the benefit of falling back to normal delete behavior is to prevent `kubectl drain` command from timing out.

Reply:
In both cases it'll fall back to delete. It will fall back if it can't evict after some timeout. I think you may want to remove the conditional and always delete as I'm seeing issues (see the above comment) evicting longhorn pods when upgrading kubernetes as it requires us to roll the entire cluster.

Reply:
After discussing IRL we agreed that the original approach from @rrpolanco was best.